query-algebrizer/src/clauses/mod.rs

```diff
@@ -240,6 +243,38 @@ impl ConjoiningClauses {
             ..Default::default()
         }
     }
+
+    pub fn with_inputs<T>(in_variables: BTreeSet<Variable>, inputs: T) -> ConjoiningClauses
+    where T: Into<Option<QueryInputs>> {
+        ConjoiningClauses::with_inputs_and_alias_counter(in_variables, inputs, RcCounter::new())
+    }
+
+    pub fn with_inputs_and_alias_counter<T>(in_variables: BTreeSet<Variable>,
+                                            inputs: T,
+                                            alias_counter: RcCounter) -> ConjoiningClauses
+    where T: Into<Option<QueryInputs>> {
+        match inputs.into() {
+            None => ConjoiningClauses::with_alias_counter(alias_counter),
+            Some(QueryInputs { mut types, mut values }) => {
+                // Discard any bindings not mentioned in our :in clause.
```

Comment:
Do we test somewhere that all `:in` variables appear?  I feel like Datomic probably doesn't care, but we should probably fail in this case, since it's likely to be an error.